ubsip_vhci: locking in irp processing cleaned up

bartlomiejcieszkowski opened the pull request:
This cleans up the locking and handling of irps in usbip_vhci

Also the irps for device ioctl were coming at dpc which caused a lot of pain in handling - changed to process these in a thread - see threaded_csq.[c|h] - a nice wrapper

Tested this with an xbox one controller - as per #122 
Did the whole attach / detach scenario, everything works great

Also did some stress testing with following scenario:

1. Attached 3 devices:
* Microsoft Xbox One controller
* Logitech usb mouse
* Pendrive

2. Pressed buttons, copied 2 gigabyte iso from pendrive
3. Detached with `usbip detach -p 0`

Result: no BSOD

Also it is worth noting that it's safe to reduce SleepEx in usbip_forward.c to 100ms from 500ms

bartlomiejcieszkowski commented:
in the meantime ill cleanup the USING_CSQ(...) define as this is some leftover

bartlomiejcieszkowski commented on driver/vhci/vhci_ioctl.c:
```diff
@@ -133,7 +142,22 @@ setup_topology_address(pusbip_vpdo_dev_t vpdo, PIO_STACK_LOCATION irpStack)
 	return STATUS_SUCCESS;
 }
 
-PAGEABLE NTSTATUS
```
worth noting: removed PAGEABLE from vhci_internal_ioctl, as this was called from DPC

cezanne replied:
vhci_internal_ioctl(IRP_MJ_INTERNAL_DEVICE_CONTROL) is called at IRQL=PASSIVE_LEVEL. PAGEABLE code would be OK.

bartlomiejcieszkowski replied:
it's not actually, found out during debug as i added PAGED_CODE() inside and this one is called at dispatch, so the pageable actually could have resulted in page fault if this was paged out

cezanne replied:
Well, I have checked this routine is called at IRQL=DISPATCH_LEVEL. I agreed that this change should be applied.

koudis commented:
### Review
The technical concern must review @cezanne.

### Test
I will/can test branch with 
- Yubico,
- USB 3.1 Disk
- multiple Unifying devices - keyboad and mouse

### SleepEx

The problem with SleepEx is solved in [forwarder](https://github.com/cezanne/usbip-win/tree/forwarder) branch which deprecates #119

cezanne commented:
@bartlomiejcieszkowski : 
> Also the irps for device ioctl were coming at dpc which caused a lot of pain in handling - changed to process these in a thread

Please, can you explain DPC pain ❓  In a usual case, thread will be more painful than DPC.
In my opinion, if a DPC routine is called with an unnecessary lock or something, the DPC routine itself should be fixed.

bartlomiejcieszkowski commented:
@cezanne nevermind that comment, i've rechecked the patch with different paths taken - and removed CSQ altogether, as it deemed unnecessary :)

I've explored some paths like
1) CSQ
2) CSQ with passive level lock for urbrs (GuardedMutex) - oh that was fun :D .. but was killed by DPC latency introduced by network adapter :(
3) No CSQ

Yeah and the no csq version works, so i'm dropping it as it is unnecessary, i've simplified the pull request.

cezanne commented on driver/vhci/usbreq.c:
```diff
@@ -104,8 +104,8 @@ static void
 {
 	struct urb_req	*urbr;
-
-	KeAcquireSpinLockAtDpcLevel(&vpdo->lock_urbr);
```
The previous code would suffice. Because [MSDN](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-iosetcancelroutine) says
> Driver Cancel routines are called at IRQL = DISPATCH_LEVEL with the cancel spin lock held.

bartlomiejcieszkowski replied:
i'm releasing early CancelSpinlock as we don't need it, this is for os responsivnes.

From [MSDN](https://docs.microsoft.com/pl-pl/windows-hardware/drivers/ddi/wdm/nc-wdm-driver_cancel):
> The I/O manager calls IoAcquireCancelSpinLock before calling a driver's Cancel routine, so the Cancel routine must call IoReleaseCancelSpinLock at some point. The routine should not hold the spin lock longer than necessary.

So to be nice for the rest of drivers we release it asap - multicore.

cezanne replied:
`KeAcuiqreSpinLockAtDpcLevel` may be faster than `KeAcquireSpinLock()`. My comment meant that `KeAcquireSpinLockAtDpcLevel` should be used instead  of `KeAcquireSpinLock()` because `remove_cancelled_urbr` is already at DISPATCH_LEVEL.

bartlomiejcieszkowski replied:
yeah, i'm aware, but with this change it is no more at the dpc level at this point in a code

cezanne replied:
Which routine calls`remove_cancelled_urbr` with a lower IRQL than DISPATCH?

bartlomiejcieszkowski replied:
it's not much about calling irql lower than dispatch, it's also about locking order
we have 2 possible locks:
A) lock_urbr
B) global cancel lock

lets say we have functions acquire - a, release - r

Throughout the code we call
a(A) -> a(B) -> (...) -> r(B) -> r(A)

so actually in the cancel routine if we were holding lock B and acquiring A we would be doing
(from cancel routine caller) a(B) -> a(A) -> r(A) -> r(B)

this might have resulted in a deadlock if we had on different processor A locked,
so we are doing in this patchset:
(from cancel routine caller) a(B) -> r(B) -> a(A) -> r(A)

removing the potential deadlock :)

cezanne replied:
I'm just saying once again that `KeAcquireSpinLockAtDpcLevel` should be used instead of `KeAcquireSpinLock()` because `remove_cancelled_urbr` is already at DISPATCH_LEVEL. 😢 Please remind that our long discussion is confined to code line 107-108. Of course, your deadlock comment should be considered later to review below codes.

bartlomiejcieszkowski replied:
this method is called in this patchset at usbreq.c:142:
```
	IoReleaseCancelSpinLock(irp->CancelIrql);

	remove_cancelled_urbr(vpdo, irp);
```
we are releasing cancel spinlock and we are no longer at passive dispatch_level so this is okay :)

cezanne replied:
I got it finally.. :smiley: Sorry for long discussion.

cezanne commented on driver/vhci/vhci_dev.h:
```diff
@@ -128,6 +128,7 @@ typedef struct
 	LONG	InterfaceRefCount;
 	// a pending irp when no urb is requested
 	PIRP	pending_read_irp;
+	BOOLEAN	pending_read_irp_cancellable;
```
Is this variable required? It is used only in `complete_pending_read_irp()`. It would be no problem even though  `complete_pending_read_irp()` always clears a cancel routine irregardless of that variable. Because the pending read irp will be completed right after clearing. And there seems to be no possibility that a cancel routine other than `on_pending_irp_read_cancelled` intervenes. A pending read irp stays in a vhci driver.

bartlomiejcieszkowski replied:
removed, unneeded

bartlomiejcieszkowski commented on driver/vhci/vhci_write.c:
```diff
@@ -351,21 +351,28 @@ process_write_irp(pusbip_vpdo_dev_t vpdo, PIRP irp)
+		if (valid_irp) {
+			irp->IoStatus.Status = status;
+
+			/* it seems windows client usb driver will think
```
@cezanne i must say that i tought it was weird that there is a need to raiseirql here, and had to see it myself - wow - looks like some ancient f-up in the microsoft stack

cezanne replied:
IRQL raising code was a legacy. I remember it was very weird. I'll also investigate.

koudis commented:
@bartlomiejcieszkowski  Is the branch prepared for testing?

bartlomiejcieszkowski commented:
@koudis and now it is, sorry for delay, had a busy week :)

koudis commented:
@bartlomiejcieszkowski I just test the branch. It works really nice! No lags, no BSOD. 3 devices connected at once and no problem.

I will continue tomorrow because i forgot my USB Hub at work - with 7 SSD USB 3.1 Disks (i've got BSOD with 50% probablity with old USBIP vhci driver)

cezanne commented:
@bartlomiejcieszkowski : Please apply [this patch](https://github.com/cezanne/usbip-win/files/4515759/beautify.diff.gz) for beautify.

bartlomiejcieszkowski commented:
@cezanne applied, squash and merge?

koudis commented:
@bartlomiejcieszkowski please wait a minute :). I will test new changes and then we can merge.

bartlomiejcieszkowski commented:
@koudis sure :)

koudis commented:
@cezanne @bartlomiejcieszkowski I test the branch after rebase. It seems to work without problems.

I test simultaneously 3x USB 3.1 SSD drives, 1x USB 2.0 flash disk and 2x Logitech unyfing receiver. I copy file from each device to computer (simultaneously) and during "copy" I detach usb devices by `usbip.exe detach -p 0` - no problem occurred.

cezanne commented:
@bartlomiejcieszkowski :
> @cezanne applied, squash and merge?

Please squash, rebase, fast-forward merge. 😃 
Great thanks to @bartlomiejcieszkowski and @koudis .
